Add new GauntFactor class and add Itoh integrated Gaunt factors to f-f radiative loss

jwreep opened the pull request:
Fixes #248 

- [x] Parse the new files
- [x] Implement new GauntFactor class
- [x] Add Itoh options to total f-f Gaunt factor
- [x] Add Itoh options to f-f radiative loss function
- [x] Tests

jwreep commented:
For my own reference, the equations from Itoh et al 2002 for the two forms of the Gaunt factor.  (They're using log base 10.)

Relativistic:
<img width="441" alt="Screenshot 2024-08-15 at 11 42 37 AM" src="https://github.com/user-attachments/assets/73c06216-0f37-4cfc-b794-5b0b30e31767">

Non-relativistic:
<img width="434" alt="Screenshot 2024-08-15 at 11 42 56 AM" src="https://github.com/user-attachments/assets/d31af80e-5295-4059-bcd2-a3349ac61c91">

wtbarnes commented:
As we implement this, I'd like to think carefully about how these are related to the other Gaunt factor methods. Currently, we have the following methods for calculating Gaunt factors:

- `gaunt_factor_free_free`--function of wavelength and temperature
- `_gaunt_factor_free_free_itoh`
  - used by `gaunt_factor_free_free` in the relativistic regime
- `_gaunt_factor_free_free_sutherland`
  - used by `gaunt_factor_free_free` in the non-relativistic regime
- `_gaunt_factor_free_free_total`
  - function of temperature
  - wavelength-integrated version of `_gaunt_factor_free_free_sutherland` (I think)
  - used by `free_free_radiative_loss`
- `_gaunt_factor_free_bound_total`
  - function of temperature
  - used by `free_bound_radiative_loss`

In our current organization of the API, even though most of these are private methods, it can be confusing why we have so many methods named very similar things and why/where each is relevant. While implementing these new methods, whether there is some better top level API for calculating these such that we minimize the number of entry points for calculating a Gaunt factor and make it more obvious to a user (or future developer, including ourselves, why method is preferred over and another).

jwreep commented:
I'll spend a bit of time thinking about this.  Open to suggestions!

jwreep commented:
Unless I'm missing something, the actual data that's read in by `fiasco.io` for every type of Gaunt factor does not depend on the properties of an ion.  That is, the fitting coefficients are the same for all ions.  

If that's correct, perhaps it would be better to create a new module or class just dedicated to the Gaunt factor, and pull these functions out of the ion class.  Then, call some overarching function along the lines of `gaunt_factor(free_free=True, free_bound=False, wavelength_integrated=False)` that then calls the appropriate function from the class.  

This is trivial for free-free emission.  For example, `ion._gaunt_factor_free_free_itoh` doesn't depend on *any* ion properties, `ion._gaunt_factor_free_free_sutherland` only depends on charge state and thus gives the same Gaunt factors for He III, Fe III, etc. at a given temperature.  

The free-bound case is slightly more involved, because it depends on charge, ionization potential, and principal quantum number.  However, nothing about this needs to be inherently tied to the ion class and each could be passed to a function.

Would something along those lines make sense?

jwreep commented:
@wtbarnes, before I take this much further, your input would be welcome.  I've created a new `GauntFactor` class with a bit of logic that should hopefully be readable code.  It doesn't (shouldn't) affect anything outward-facing, but it does nest the various functions that calculate the different Gaunt factors into one object.  Would something along these lines be a reasonable way to deal with the convoluted mess of Gaunt factors?  Or did you have something else in mind?

(It still needs error handling and debugging, of course.  It's a bit clumsy with `.gf` to get the actual values at the end as well.)

wtbarnes commented:
Sorry for neglecting this for so long!

**tl;dr I'm pro this approach overall. I have some thoughts on how exactly to implement this object. I've left a few high-level comments below and some inline comments on the code.**

At first I was a bit skeptical of having these on a separate object, but I'm now thinking this is probably a good idea. Exposed as a separate object, they can still use the same data ingestion model as `Ion` which means we don't have to change anything at the I/O layer.

This also helps to shorten the `ions.py` file which seems to be ever expanding. 

> Edit: also, I was wrong in my previous comment that some of the Gaunt factors don't depend on any ion properties. The Ion._itoh data is attached to an ion, though some of the other data is not.

We could always change this in the way the parsing is done to be agnostic of `Ion` and then select the row appropriate to the given ion when doing the necessary calculation in the gaunt factor. I think it is per element anyway so it doesn't even need the granularity of per ion.

wtbarnes commented on fiasco/gaunt.py:
```diff
@@ -0,0 +1,231 @@
+                # These terms are eventually set to zero anyway since the ionization fraction is so small
+                # at these temperatures.
+                with np.errstate(over='ignore'):
+                    f_2 = g_n_0 * ion._zeta_0 * (z_0**4 / n_0**5) * np.exp((prefactor * z_0**2)/(n_0**2))
```
`zeta_0` should live on this object too. I don't think it is used anywhere else?

jwreep replied:
It's not used elsewhere, and is just a function of atomic number and charge.  We could move it.

jwreep commented on fiasco/gaunt.py:
```diff
@@ -0,0 +1,231 @@
+    """
+    Class for calculating the Gaunt factor for various continuum processes.
+    """
+    def __init__(self, ion, wavelength=None, freefree=False, freebound=False, 
+                    wavelength_integrated=False,
+                    hdf5_dbase_root=None, *args, **kwargs):
```
No arguments.  Part of the reason I did this was because of the `itoh` data, but I can move that.

jwreep commented on fiasco/gaunt.py:
```diff
@@ -0,0 +1,231 @@
+            self.hdf5_dbase_root = hdf5_dbase_root
+        check_database(self.hdf5_dbase_root, **kwargs)
+        
+        self.freefree = freefree
+        self.freebound = freebound
+        self.wavelength_integrated = wavelength_integrated
+        
+        self.gf = None
+        if freefree and wavelength is not None and not wavelength_integrated:
+            self.gf = self.free_free(ion, wavelength)
+        elif freefree and wavelength_integrated:
+            self.gf = self.free_free_total(ion)
+        elif freebound and wavelength_integrated:
+            self.gf = self.free_bound_total(ion, **kwargs)
```
I was thinking that it should calculate the Gaunt factor on instantiation, basically creating an object when called.  I guess you're suggesting creating an object at the same time time as the `Ion` object, and then call its methods when needed?  That's probably better.

wtbarnes replied:
I think this would be more efficient. Otherwise, a new object has to be created for every new calculation.

jwreep commented on fiasco/gaunt.py:
```diff
@@ -0,0 +1,231 @@
+        data_path = '/'.join(['continuum', 'itohintnonrel'])
+        return DataIndexer.create_indexer(self.hdf5_dbase_root, data_path)
+        
+    @u.quantity_input
+    def free_free(self, ion, wavelength: u.angstrom) -> u.dimensionless_unscaled:
+        r"""
+        Free-free Gaunt factor as a function of wavelength.
+
+        The free-free Gaunt factor is calculated from a lookup table of temperature averaged
+        free-free Gaunt factors from Table 2 of :cite:t:`sutherland_accurate_1998` as a function
+        of :math:`\log{\gamma^2},\log{u}`, where :math:`\gamma^2=Z^2\mathrm{Ry}/k_BT`
+        and :math:`u=hc/\lambda k_BT`.
+
+        For the regime, :math:`6<\log_{10}(T)< 8.5` and :math:`-4<\log_{10}(u)<1`, the above
+        prescription is replaced with the fitting formula of :cite:t:`itoh_relativistic_2000`
+        for the relativistic free-free Gaunt factor. This is given by Eq. 4 of
+        :cite:t:`itoh_relativistic_2000`,
+
+        .. math::
+
+            g_{ff} = \sum_{i,j=0}^{10} a_{ij}t^iU^j,
+
+        where :math:`t=(\log{T} - 7.25)/1.25` and :math:`U=(\log{u} + 1.5)/2.5`.
+        """
+        gf_itoh = self._free_free_itoh(ion, wavelength)
+        gf_sutherland = self._free_free_sutherland(ion, wavelength)
+        gf = np.where(np.isnan(gf_itoh), gf_sutherland, gf_itoh)
+
+        return gf
```
What's your thinking?  I was hoping to hide all the Gaunt factors from the `Ion` object.

wtbarnes replied:
That probably makes the most sense and is the most consistent approach.

jwreep commented:
Thanks for the suggestions!  I'll try to flesh this out a bit more.

jwreep commented on fiasco/gaunt.py:
```diff
@@ -0,0 +1,231 @@
+        if z == 0:
+            return u.Quantity(np.zeros(ion.temperature.shape))
+        else:
+            recombined = ion.previous_ion()
+            if not recombined._has_dataset('fblvl'):
+                return u.Quantity(np.zeros(ion.temperature.shape))
+            Ry = const.h * const.c * const.Ryd
+            prefactor = (Ry / (const.k_B * ion.temperature)).decompose()
+            n_0 = recombined._fblvl['n'][0]
+            if ground_state:
+                g_n_0 = 0.9 # The ground state Gaunt factor, g_fb(n_0), prescribed by Mewe et al 1986
+                z_0 = n_0 * np.sqrt(recombined.ip / Ry).decompose()
```
The existence is actually already checked in the ion object, so this is currently double-checking unnecessarily.

jwreep commented on fiasco/gaunt.py:
```diff
@@ -0,0 +1,231 @@
+    def _itohintnonrel(self):
+        data_path = '/'.join(['continuum', 'itohintnonrel'])
+        return DataIndexer.create_indexer(self.hdf5_dbase_root, data_path)
+        
```
A bit lost on how to do this.  It looks like I can't just move this over to the `GauntFactor` object and rebuild the database.  (`KeyError: 'continuum/itoh not found in /Users/reep/.fiasco/chianti_dbase.h5'`).  Is there something in the building of the HDF5 file that ties it to the ions?

wtbarnes replied:
It's because the `_itoh` dataset lives under '{atomic_symbol}/continuum/itoh'. There are two ways we could go about this:

1. (Low effort) Require Z (or some element designation) when instantiating the `GauntFactor` constructor, store it internally, and just copy the current `_itoh` property from `Ion`.
2. (Medium effort) Implement a method on `GauntFactor` that takes in an elemental symbol and returns the appropriate part of that dataset. This requires no changes to the database or parsing code, but does make this data access pattern inconsistent with the rest of the package.
3. (Most effort) Move the Itoh dataset underneath '/continuum/itoh', add the property onto `GauntFactor` in the same manner as the others. Anytime this data is needed, pull out the correct row from that table for the appropriate Z value. This will require passing in Z wherever this is needed. This will also require rebuilding the database after changing the schema.

If we chose option 1, we could also just require passing in an `Ion` object when instantiating `GauntFactor` (in contrast to what I said previously). Maybe that does actually make the most sense?

jwreep replied:
I just implemented Option 3.  Took me a bit to figure out the parser for it, but it's pulling the correct values and is now stored under `GauntFactor`.  

It makes the most sense to me, and a few of the methods need the atomic number anyway.

jwreep commented:
@wtbarnes, has something changed with the github actions in the recent pulls?  Pytest is passing on my machine, but failing on Github (including lots of tests not related to this PR).

wtbarnes commented on fiasco/ions.py:
```diff
@@ -61,6 +60,7 @@ def __init__(self, ion_name, temperature: u.K,
         self._dset_names['ioneq_filename'] = kwargs.get('ioneq_filename', 'chianti')
         self._dset_names['ip_filename'] = kwargs.get('ip_filename', 'chianti')
         self.abundance = abundance
+        self.gaunt_factor = GauntFactor()
```
```suggestion
        self.gaunt_factor = GauntFactor(hdf5_dbase_root=self.hdf5_dbase_root)
```
You need to pass the same path to the database to the gaunt factor object as is used by the ion object. The CI is failing because it's using the default option which is to look in $HOME/.fiasco but that doesn't exist and it hangs when it asks if you want to build it. The CI runner will use a temporary database built for every test run that is passed through to all of the ion object fixtures. It passes locally for you because you have a local version of the database in ~/.fiasco.

jwreep replied:
Aha!  Thanks.

jwreep commented:
Similar to the discussion in #280, I think there's a design choice we need to make here.  If someone is working with v8.0.7, then `GauntFactor.free_free_total(temperature, charge_state, itoh=True)` will fail because the data is missing, so we need to decide on error handling for that.  

The options, I think would be:

1. Simply return an error message to the user that the data doesn't exist for the pertinent version of CHIANTI.
2. Default to the Sutherland form, warning the user.
3. Default to the Sutherland form silently.

The differences between each of the three forms is relatively minor, and I've already added some logic to prioritize the Sutherland data outside the valid temperature ranges for the Itoh data.

Any thoughts?

wtbarnes commented:
Looking at how the integrated gaunt factor is calculated, I'm finding the logic confusing. I would suggest implementing this in a manner similar to what is done in `gaunt_factor_free_free`.

Commenting line-by-line isn't so useful at this point, so here's some pseudo code that outlines my thinking:

```python
def free_free_integrated():
    gf_sutherland = free_free_sutherland_integrated()
    gf_itoh = free_free_itoh_integrated()
    return where(isnan(gf_itoh), gf_sutherland, gf_itoh)

@needs_dataset('gffint')
def free_free_sutherland_integrated():
    # do sutherland calculations, data always available and always valid

def free_free_itoh_integrated():
    try:
        gf_rel = free_free_itoh_integrated_relativistic()
    except MissingDatasetException:
        gf_rel = array(NaN)
    try:
        gf_nonrel = free_free_itoh_integrated_nonrelativistic()
    except MissingDatasetException:
        gf_nonrel = array(NaN)
    return where(isnan(gf_nonrel), gf_rel, gf_nonrel)

@needs_datatset('itoh_integrated_gaunt')
def free_free_itoh_integrated_relativistic():
    # Do Itoh relativistic calculation, returning NaNs outside of valid range
    # calls no other functions

@needs_dataset('itoh_integrated_gaunt_nonrel')
def free_free_itoh_integrated_nonrelativistic():
    # Do Itoh nonrelativistic calculation, returning NaNs outside of valid range
    # calls no other functions
```

A few comments on this:
- This proposed structure separates the logic of the calculation of each form from the decision to use one or the other
- This proposed structure "tiers" the decision making to first deciding between Sutherland or Itoh and then between relativistic and non-relativistic.
- The `try...except` calls are a bit clunky. We could use something like the `_has_dataset` property on `Ion` to clean this up.
- I don't think I'm in favor of having `itoh` or `relativistic` keywords. I think these choices are already made by the presence of particular datasets. Presumably one always wants to use a relativistic approach if available in a certain energy range. Similarly and as in the non-integrated GF calculation, the Itoh approach is considered more accurate than the Sutherland approach and thus should be used if available. 
- Without the keyword argument as above, I don't think we need to warn users about the lack of data present. That is just a consequence of using a particular version of the database. I think this is closest to your option 3.
- I wonder too if we could switch at least the designation on the gaunt factors from "total" to "integrated"? Right now there is a mix of both, but I think integrated is actually more accurate?

Let me know what you think about this.

jwreep commented:
Thank you for the input.  I appreciate it.  

> This proposed structure "tiers" the decision making to first deciding between Sutherland or Itoh and then between relativistic and non-relativistic.

One reason why I chose to use keywords is simply that CHIANTI itself does not actually use the Itoh form, so I was considering this as an optional alternative.  The differences in the Gaunt factors are small, though, perhaps 1-2%.  I'm not otherwise tied to the keyword approach, so I can remove it and follow the suggested approach.  I would expect the f-f radiative loss comparisons to give a slightly larger error as a consequence, so the tolerance for those tests might need to be loosened?

In any case, I can try to rearrange the logic.

And you're right, I have been mixing and matching "total" and "integrated" so I'll unify that.

wtbarnes commented:
> One reason why I chose to use keywords is simply that CHIANTI itself does not actually use the Itoh form, so I was considering this as an optional alternative. 

Ah! I didn't fully realized that. In that case, yeah maybe the top level function could then just be modified to be (again using my pseudo code approach as above):

```python
def free_free_integrated(use_itoh=False):
    gf = free_free_sutherland_integrated()
    if use_itoh:
        gf_itoh = free_free_itoh_integrated()
        gf = where(isnan(gf_itoh), gf, gf_itoh)
    return gf
```

jwreep commented:
Yeah, I might be missing something, but the radiative loss calculations in the IDL code don't actually seem to call the Itoh version anywhere that I can see.  Presumably they intended to do so since they've included the data files, but I don't actually see where they're being used.  There's [ff_gaunt_itoh.pro](https://sohoftp.nascom.nasa.gov/solarsoft/packages/chianti/idl/continuum/ff_gaunt_itoh.pro) that reads both files and calculates the Gaunt factor, but [ff_rad_loss.pro](https://sohoftp.nascom.nasa.gov/solarsoft/packages/chianti/idl/continuum/ff_rad_loss.pro) doesn't actually call it (only the Sutherland form).  

So, perhaps functionality that's intended but hasn't been implemented yet?

wtbarnes commented:
That's odd. As far as I know, that's not part of the upcoming v11 changes. Perhaps @pryoung would know more?

jwreep commented:
To get the tests to run for version 9, I've added a new test file list and changed the logic on when to read it.  There was a note in comments that this needs to be generalized.  What I've done seems to work, but might not be the most elegant.

jwreep commented:
Marking this as ready for review, as I think the intended functionality is working.  Codecov is complaining a bit, but it seems to be for functions that **_are_** being tested in the version 9 tests, just not the version 8 tests.  Happy to add more tests if desired, though.

wtbarnes commented on fiasco/collections.py:
```diff
@@ -363,11 +365,16 @@ def bound_bound_radiative_loss(self, density, **kwargs) -> u.Unit('erg cm3 s-1')
+        Parameters
+        ----------
+        use_itoh : `bool`, optional
+            Whether to use the Itoh Gaunt Factors.  Defaults to false.
```
Would be better to include a citation to the paper here so it is explicit exactly which Itoh Gaunt factors we mean. Could also link to the appropriate method that gives more information about this calculation.

wtbarnes commented on fiasco/collections.py:
```diff
@@ -315,14 +315,16 @@ def radiative_loss(self, density: u.cm**(-3), **kwargs) -> u.Unit('erg cm3 s-1')
         """
         rad_loss_bound_bound = self.bound_bound_radiative_loss(density, **kwargs)
-        rad_loss_free_free = self.free_free_radiative_loss()
+        rad_loss_free_free = self.free_free_radiative_loss(use_itoh)
```
```suggestion
        rad_loss_free_free = self.free_free_radiative_loss(use_itoh=use_itoh)
```

wtbarnes commented on fiasco/collections.py:
```diff
@@ -315,14 +315,16 @@ def radiative_loss(self, density: u.cm**(-3), **kwargs) -> u.Unit('erg cm3 s-1')
         density : `~astropy.units.Quantity`
             Electron number density
+        use_itoh : `bool`, optional
+            Whether to use the Itoh Gaunt Factors for free-free.  Defaults to false.
```
See below

wtbarnes commented on fiasco/__init__.py:
```diff
@@ -4,6 +4,7 @@
 from fiasco.collections import IonCollection
 from fiasco.elements import Element
 from fiasco.fiasco import list_elements, list_ions, proton_electron_ratio
+from fiasco.gaunt import GauntFactor
```
Are we sure we want this to live in the top level namespace? I imagine most user interaction with this to be via the `Ion` object.

wtbarnes commented on docs/conf.py:
```diff
@@ -155,7 +155,7 @@
         ascii_dbase_url = CHIANTI_URL.format(version=LATEST_VERSION),
         ask_before=False,
-        files=get_test_file_list(),
+        files=list(),
```
See note below in `conftest.py`. Same logic here.

wtbarnes reviewed (changes requested):
Overall, I think this PR is in good shape. I've left a bunch of comments and suggestions, but most are just minor. The two big suggestions are:

- I'm confused as to why multiple test file lists are needed per version. I understand that some files will be present in some versions of the database and absent in others, but when building the HDF5 database, any missing files should just be skipped without throwing an exception. Is this not the case? I'm also not a fan of the logic of choosing versions of the file list as currently implemented (see my specific comments), but if my hunch about not actually needing multiple files is correct, then that is a moot point
- As implemented, I find the logic of calculating the Itoh (relativistic and non-relativistic) GF quite confusing. I also am wary of how performance will fare for large arrays of temperatures. I've offered some suggestions for refactoring. I think these should work as suggested, but I've not tested them.

As always, thanks for this massive effort and for your patience!

jwreep commented:
Thanks, Will!  I'll try to get all of this cleaned up over the coming week.  

-- I was having trouble getting the tests to actually pass with just one test file list.  Could be entirely that I was doing something stupid, but I had tried adding the new files to the original list and the tests were failing to find the files, then.  Not sure why -- and entirely possible it was my own mistake.  I'll try it again to see if I can fix it.

-- The Einstein summation was considerably faster than looping over the matrix itself, which was in my original implementation.  I still had nested it under a temperature loop, though, because I was trying to pull out the invalid temperatures.  Poor design on my part.  I'll revisit this.

pryoung commented:
Hi Will, the Itoh f-f Gaunt factors were added a long time ago (20 years?) to CHIANTI. I recall there has been a more recent calculation since that we should put in CHIANTI at some point. Peter


From: Will Barnes ***@***.***>
Date: Friday, September 20, 2024 at 7:38 PM
To: wtbarnes/fiasco ***@***.***>
Cc: Young, Peter R. (GSFC-6710) ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] [BULK] Re: [wtbarnes/fiasco] Add new GauntFactor class and add Itoh integrated Gaunt factors to f-f radiative loss (PR #293)
CAUTION: This email originated from outside of NASA.  Please take care when clicking links or opening attachments.  Use the "Report Message" button to report suspicious messages to the NASA SOC.



That's odd. As far as I know, that's not part of the upcoming v11 changes. Perhaps @pryoung<https://github.com/pryoung> would know more?

—
Reply to this email directly, view it on GitHub<https://github.com/wtbarnes/fiasco/pull/293#issuecomment-2364753194>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEL4UCFANSDB5ZXOVBYYIRDZXSWVNAVCNFSM6AAAAABMS7IS3WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNRUG42TGMJZGQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

wtbarnes commented:
> Thanks, Will! I'll try to get all of this cleaned up over the coming week.
> 
> -- I was having trouble getting the tests to actually pass with just one test file list. Could be entirely that I was doing something stupid, but I had tried adding the new files to the original list and the tests were failing to find the files, then. Not sure why -- and entirely possible it was my own mistake. I'll try it again to see if I can fix it.

Interesting. It could be that I'm missing something. What exception are you seeing specifically? There could be some weirdness in the test setup that is causing this.
 
> -- The Einstein summation was considerably faster than looping over the matrix itself, which was in my original implementation. I still had nested it under a temperature loop, though, because I was trying to pull out the invalid temperatures. Poor design on my part. I'll revisit this.

Ah I see. Yes, I'm not surprised that the Einstein summation makes things faster if you're already looping over temperature. I think my proposed solution should be more performant, even with the loop over the coefficients.

wtbarnes commented:
> Hi Will, the Itoh f-f Gaunt factors were added a long time ago (20 years?) to CHIANTI. I recall there has been a more recent calculation since that we should put in CHIANTI at some point. Peter 

Thanks for the clarification Peter. Are you referring to the integrated free-free Itoh Gaunt factors (from Itoh et al. 2002) or the wavelength resolved Gaunt factors (from Itoh et al. 2000)? For clarification, I was talking about the former (the integrated ones). I know that the later are used in calculating the free-free continuum as a function of wavelength and temperature in CHIANTI. However, it seems that the integrated Itoh 2002 Gaunt factors (relativistic and non-relativistic) are not actually used in the free-free contribution to the wavelength-integrated radiative losses. Was there a reason for that?

jwreep commented on fiasco/io/sources/continuum_sources.py:
```diff
@@ -9,8 +9,9 @@
 
-__all__ = ['GffguParser', 'GffintParser', 'KlgfbParser', 'VernerParser', 'ItohParser',
-           'HSeqParser', 'HeSeqParser']
+__all__ = ['GffguParser', 'GffintParser', 'ItohIntegratedGauntParser',
+           'ItohIntegratedGauntNonrelParser', 'KlgfbParser', 'VernerParser',
+           'ItohParser', 'HSeqParser', 'HeSeqParser']
```
¯\_(ツ)_/¯

jwreep commented on fiasco/gaunt.py:
```diff
@@ -0,0 +1,410 @@
+
+    @u.quantity_input
+    def free_free_integrated(self, temperature: u.K, charge_state, use_itoh=False) -> u.dimensionless_unscaled:
+        """
+        The wavelength-integrated free-free Gaunt factor, used for calculating
+        the total radiative losses from free-free emission.
+
+        Parameters
+        ----------
+        temperature : `~astropy.units.Quantity`
+            The temperature(s) for which to calculate the Gaunt factor
+        charge_state : `int`,
+            The charge state of the ion
+        use_itoh : `bool`, optional
+            Whether to use the Itoh Gaunt Factors.  Defaults to false.
+        """
```
I guess I have a question about design.  I realize that the private ones are not parsed by the docs, but isn't it still useful to keep this information for people viewing the code?

wtbarnes replied:
Yeah, there's certainly no harm in keeping the private ones as it can make the code more readable and add more context.

pryoung commented:
OK, yes, we went for the Sutherland data for the integrated losses. See https://www.chiantidatabase.org/tech_reports/09_rad_loss/chianti_report_09.pdf (Sect. 3.1). This was because the temperature coverage for Itoh wasn’t as wide as Sutherland, and they agree very well with each other (where they overlap).

                Peter


From: Will Barnes ***@***.***>
Date: Monday, September 23, 2024 at 3:40 PM
To: wtbarnes/fiasco ***@***.***>
Cc: Young, Peter R. (GSFC-6710) ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] [BULK] Re: [wtbarnes/fiasco] Add new GauntFactor class and add Itoh integrated Gaunt factors to f-f radiative loss (PR #293)
CAUTION: This email originated from outside of NASA.  Please take care when clicking links or opening attachments.  Use the "Report Message" button to report suspicious messages to the NASA SOC.



Hi Will, the Itoh f-f Gaunt factors were added a long time ago (20 years?) to CHIANTI. I recall there has been a more recent calculation since that we should put in CHIANTI at some point. Peter

Thanks for the clarification Peter. Are you referring to the integrated free-free Itoh Gaunt factors (from Itoh et al. 2002) or the wavelength resolved Gaunt factors (from Itoh et al. 2000)? For clarification, I was talking about the former (the integrated ones). I know that the later are used in calculating the free-free continuum as a function of wavelength and temperature in CHIANTI. However, it seems that the integrated Itoh 2002 Gaunt factors (relativistic and non-relativistic) are not actually used in the free-free contribution to the wavelength-integrated radiative losses. Was there a reason for that?

—
Reply to this email directly, view it on GitHub<https://github.com/wtbarnes/fiasco/pull/293#issuecomment-2369208239>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEL4UCBOPE3UA4XJN3QS6CTZYBVBHAVCNFSM6AAAAABMS7IS3WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNRZGIYDQMRTHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

jwreep commented on fiasco/io/sources/continuum_sources.py:
```diff
@@ -93,6 +94,112 @@ def __init__(self, filename, **kwargs):
+
+    def __init__(self, filename, **kwargs):
+        super().__init__(filename, **kwargs)
+        self.full_path = pathlib.Path(kwargs.get('full_path',
+                                    self.ascii_dbase_root / 'continuum' / filename))
```
Is this true for all of the parsers?  This function is more or less the same for all of them.

wtbarnes replied:
Huh. Yeah, you're right. Honestly, I have no idea why `full_path` is treated as a possible kwarg in so many of the continuum and non-ion source parsers. I suspect it should probably be removed everywhere, but that's not in scope for this PR. Feel free to do whatever here and we can sort it out elsewhere.

jwreep replied:
I'll leave it for now.  The coding for the parsers is Fortran to me.

jwreep commented on fiasco/gaunt.py:
```diff
@@ -0,0 +1,410 @@
+        index_nl = np.where(np.logical_and(self._klgfb['n'] == n, self._klgfb['l'] == l))[0]
+        # If there is no Gaunt factor for n, l, set it to 1
+        if index_nl.shape == (0,):
+            gf = 1
```
Yeah, probably should be an array if there's an array of scaled energies given to the function.

jwreep commented:
Thank you for the clarification, Peter!  

I'll add a citation to this technical report in fiasco.

jwreep commented on fiasco/conftest.py:
```diff
@@ -47,7 +41,7 @@ def hdf5_dbase_root(ascii_dbase_tree, tmpdir_factory, request):
         test_files = None
     else:
-        test_files = TEST_FILES
+        test_files = list() # empty list only to signify not None
```
Let me revisit this.  I had issues because it's calling another fixture.

wtbarnes replied:
Understandable. The test setup is pretty convoluted here.

wtbarnes commented:
Looking at the docs, I now realize that the API docs for the new `GauntFactor` object aren't rendered anywhere because I asked for it to be excluded from the top-level namespace. It probably makes sense to just add it back to the top-level namespace. Sorry!

jwreep commented:
Yeah, that probably explains why the docs build failed.  lol

wtbarnes commented:
I also take back what I said about only needing this in the context of the Ion object. It's not unreasonable one might want to calculate a Gaunt factor for other reasons as well so making the API more visible is only a good thing.

jwreep commented on fiasco/util/data/test_file_list_v9.0.1.json:
How do I create the hash table?  This might be the source of my issues.

wtbarnes replied:
There's a script in `tools` in the top level directory you can use to generate it. So if you're in the root of the repo, run,

```shell
python tools/generate_hash_table.py --database /path/to/chianti/ascii/dbase
```

where "dbase" should be the top of the CHIANTI database tree. It should then generate a new hash table for your version in `fiasco/util/data`.

jwreep replied:
That seems to have worked.  I've most of the changes, and it's passing pytests for different database versions in the terminal.  Hopefully the Github tests pass, too.

jwreep commented on fiasco/tests/test_ion.py:
```diff
@@ -346,52 +339,42 @@ def test_free_free(ion):
-
-def test_free_free_radiative_loss(h1, fe20):
+@pytest.mark.parametrize(('index','expected'),
+                        [(0, 1.79093013e-22),
+                        (75, 3.06577525e-21), ])
+def test_free_free_radiative_loss(h1, fe20, index, expected):
     assert fe20.free_free_radiative_loss().shape == fe20.temperature.shape
     assert u.allclose(h1.free_free_radiative_loss(), 0.0 * u.erg * u.cm**3 / u.s)
     # This value has not been tested for correctness
-    assert u.isclose(fe20.free_free_radiative_loss()[0], 1.79093013e-22 * u.erg * u.cm**3 / u.s)
+    assert u.isclose(fe20.free_free_radiative_loss()[index], expected * u.erg * u.cm**3 / u.s)
+
+@pytest.mark.requires_dbase_version('>= 9.0.1')
+@pytest.mark.parametrize(('index','expected'),
+                        [(0, 1.79093013e-22),
+                        (75, 3.05628894e-21), ])
+def test_free_free_radiative_loss_itoh(h1, fe20, index, expected):
+    """
+    For database versions >= 9.0.1, the Itoh Gaunt factors give a different
+    result for the free-free radiative loss at temperature > 1 MK.
+    """
+    assert fe20.free_free_radiative_loss(use_itoh=True).shape == fe20.temperature.shape
+    assert u.allclose(h1.free_free_radiative_loss(use_itoh=True), 0.0 * u.erg * u.cm**3 / u.s)
+    # This value has not been tested for correctness
+    assert u.isclose(fe20.free_free_radiative_loss(use_itoh=True)[index], expected * u.erg * u.cm**3 / u.s)
```
Not sure about that.  The value returned by `fe20.free_free_radiative_loss(use_itoh=True)[75]` depends on the database version.  For database versions < 9.0.1, it will give the Sutherland result, and for > 9.0.1 it will give the Itoh result, which is slightly different.

jwreep replied:
In other words, if I just parametrized with `use_itoh`, the pytest will fail for either the newer or older versions.  I would need to parametrize on the database version -- is that possible?

wtbarnes replied:
Oops I missed the `requires_dbase_version` mark on the second test. There might be a way to combine them with that mark somehow, but it would probably be overly convoluted. Leaving them separate is fine. Sorry for the confusion!

jwreep commented:
I think that should've addressed all of your comments!  Let me know if there are any more changes you'd like.

wtbarnes commented:
> I think that should've addressed all of your comments! Let me know if there are any more changes you'd like.

Everything is looking pretty good to me! i just pushed a few changes mostly cleaning up the docstrings, linking between methods, and enforcing some consistency between a few of the equations. Have a look over the rendered docs for `GauntFactor` and relevant methods on `Ion` and let me know what you think. I think this should be basically done.

jwreep commented:
Looks fine to me.  Thanks for touching that up.

jwreep commented:
Thanks for the help!  This ended up being a lot more involved than just parsing two new data sets.  Hopefully the new format is easy to maintain and extend.  For example, the CHIANTI technical report mentions a newer data set that's more accurate, that could potentially be readily built in to fiasco.
